Feature remove elements

ililikakis opened the pull request:
**_Describe your changes here:_**



**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

- [x] The author added a BSD statement to `doc/` (or already has one)
- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)
- [ ] New source/header files are properly added to the Makefiles
- [ ] New Datatypes are added to t8indent_custom_datatypes.txt
- [ ] The reviewer executed the new code features at least once and checked the results manually
- [ ] The code is covered in an existing or new test case
- [ ] New tests use the Google Test framework
- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)
- [ ] Testing of this template: If you feel something is missing from this list, contact the developers

ililikakis committed:
Set num_elements in one step with is_family=0

ililikakis committed:
Added remove element case in t8_forest_adapt_refine_recursive

ililikakis committed:
Added level_max for recursive refinement

ililikakis committed:
Added mpi_rank as condition for is_family-test on boundery

ililikakis committed:
Added some temp debug help

ililikakis committed:
Added test for is_family in boundary check for parallelization

ililikakis committed:
Bugfix in test3 for is_family

ililikakis committed:
Added new test

ililikakis committed:
Expanding test

ililikakis committed:
Merge from develop

ililikakis committed:
Added new include

ililikakis committed:
Added flag in forest for (in)completeness

ililikakis committed:
Added test for overlapping boundery elements on prozess boarder

ililikakis committed:
Adapted function for calculating tree_offsets in incomplete forests

ililikakis committed:
Added check for overlapping elements on prozess

ililikakis committed:
Using t8_element_nca instead of several t8_element_parent

ililikakis committed:
Added debugf

ililikakis committed:
Adapted tests

ililikakis committed:
Merge remote-tracking branch 'origin/develop' into feature-remove-elements

ililikakis committed:
Removed include <t8_forest_vtk.h>

ililikakis committed:
Extended t8_forest_iterate_replace for incomplete forest

ililikakis committed:
Add test for t8_forest_iterate_replace()

ililikakis committed:
Adapted some tests

ililikakis committed:
Add example moving circle

ililikakis committed:
updated test

ililikakis committed:
fixed merge conflicts, but no bugfix

ililikakis committed:
num_children/num_siblings conflict solved

ililikakis committed:
Extended test to pyramides

ililikakis committed:
Merge branch 'develop' into feature-remove-elements

ililikakis committed:
Added first test for interpolation

ililikakis committed:
Merge branch 'develop' into feature-remove-elements

ililikakis committed:
extra test 2&3 as long t8_element_nca is not working for pyramids

ililikakis committed:
Extended test for removing elements

ililikakis committed:
cleaning

ililikakis committed:
added debugging for calculating NCA

ililikakis committed:
replaced t8_element_nca with _parent and _compare

ililikakis committed:
added new function t8_forest_is_incomplete_family

ililikakis committed:
Merge branch 'feature-remove-elements' of https://github.com/ililikakis/t8code into feature-remove-elements

ililikakis committed:
Slimmed down tests for incomplete families.

ililikakis committed:
Played around with tests

ililikakis committed:
fixed issue because of pointer

ililikakis committed:
playing around with tests

ililikakis committed:
Added new function t8_forest_remove_tree

ililikakis committed:
fixed treecounter issue

ililikakis committed:
Added documentation

ililikakis committed:
Fixed first/last_local_tree issue and added documentation

ililikakis committed:
Added new test for recursive refine and coarsening

ililikakis committed:
forest not has to be commited any more

ililikakis committed:
Added first try of recursive coarsening for incomplete forests

ililikakis committed:
Added tree removing after tree-loop

ililikakis committed:
Played around with tests

ililikakis committed:
fixed communication issue with empty trees in t8_forest_tree_shared

ililikakis committed:
Added T8_ASSERT

ililikakis committed:
Fixed spelling

ililikakis committed:
extended documentation

ililikakis committed:
Get the element_last via t8_forest_element_in_tree

ililikakis committed:
Get the elements via t8_forest_element_in_tree

ililikakis committed:
Fixed memory leak

ililikakis committed:
Changed MPI_INT64_T to T8_MPI_GLOIDX

ililikakis committed:
Fixed spelling

ililikakis committed:
renamed variable

ililikakis committed:
renamed variables

ililikakis committed:
rearranged DEBUG wrapper around tests for overlapping elements

ililikakis committed:
rearranged DEBUG wrapper around tests for overlapping elements

ililikakis committed:
reorder removing trees

ililikakis committed:
fixed recursive coarsening

ililikakis committed:
fixed mpi and seg vault issue

ililikakis committed:
played around with examples

ililikakis committed:
added sketch of benchmark for incomplete forests

ililikakis committed:
Merge branch 'main' into feature-remove-elements

ililikakis committed:
added pyramides to test

ililikakis committed:
removed old unused methodes for test of incomplete family

ililikakis committed:
renamed methode

ililikakis committed:
reordered initial variables

ililikakis committed:
fixed spelling

ililikakis committed:
fixed T8_ASSERT issue

ililikakis committed:
played around with examples

ililikakis committed:
removed unused methode

ililikakis committed:
extended documentation

ililikakis committed:
added is_famiy test in forest_adapt.cxx

ililikakis committed:
cleanup t8_tests_remove_rand

ililikakis committed:
Extended documentation.

ililikakis committed:
clean up

ililikakis committed:
Deleted old testfile

ililikakis committed:
added temp files to construct figures

ililikakis committed:
fixed latency issue in MPI communication

ililikakis committed:
added examples for thesis

ililikakis committed:
added benchmark

ililikakis committed:
added new benchmark carved balls

ililikakis committed:
update benchmarks

ililikakis committed:
update benchmarks

ililikakis committed:
update interpolation

ililikakis committed:
update benchmarks

ililikakis committed:
added benchmark

ililikakis committed:
fixed MAkefile

ililikakis committed:
update is_incomplete_family

ililikakis committed:
update benchmark sponge

ililikakis committed:
fixed timing in benchmark sponge

ililikakis committed:
fix in benchmark sponge

ililikakis committed:
update benchmark sponge

ililikakis committed:
reordered callback

ililikakis committed:
update stuff benchmarks

ililikakis committed:
update benchmark

ililikakis committed:
update benchmark

ililikakis committed:
update benchmark

ililikakis committed:
update benchmark

ililikakis committed:
merged from main

ililikakis committed:
reorderd example, benchmark and test files

Davknapp commented on example/figure_thesis/readme.txt:
```diff
@@ -0,0 +1 @@
+This folder and its content is temporary and will be deleted after completion of Ioannis thesis. The content is used to create images.
```
If you do not intend to add this example to t8code, then do not add it to the repository. For reproducability you can add it to [t8data](https://github.com/DLR-AMR/t8data).

Davknapp commented on example/Makefile.am:
```diff
@@ -12,3 +12,6 @@ include example/geometry/Makefile.am
 include example/version/Makefile.am
+
+include example/remove/Makefile.am
+include example/figure_thesis/Makefile.am
```
For reproducability you can add the figure_thesis folder to [t8data](https://github.com/DLR-AMR/t8data)

Davknapp commented on example/figure_thesis/Makefile.am:
```diff
@@ -0,0 +1,9 @@
+	example/figure_thesis/t8_ring
+
+example_figure_thesis_t8_ring_SOURCES = \
+	example/figure_thesis/t8_ring.cxx
```
For reproducability you can add the thesis example to [t8data](https://github.com/DLR-AMR/t8data)

Davknapp commented on example/figure_thesis/t8_is_family_fail.cxx:
```diff
@@ -0,0 +1,109 @@
+#include <t8_forest.h>
+#include <t8_schemes/t8_default/t8_default_cxx.hxx>
+#include <t8_vec.h>
+
```
Plaese add a description what this example demonstrates.

Davknapp commented on example/figure_thesis/t8_is_family_fail.cxx:
```diff
@@ -0,0 +1,109 @@
+  };
+
+
+int
```
Please add documentation

Davknapp commented on example/figure_thesis/t8_is_family_fail.cxx:
```diff
@@ -0,0 +1,109 @@
+  const struct t8_adapt_data *adapt_data = (const struct t8_adapt_data *) t8_forest_get_user_data (forest);
+  
+  double  centroid[3];
+  double  dist;
```
```suggestion
```
Declare it later and make it `const`

Davknapp commented on example/figure_thesis/t8_is_family_fail.cxx:
```diff
@@ -0,0 +1,109 @@
+  T8_ASSERT (adapt_data != NULL);
+
+  t8_forest_element_centroid (forest_from, which_tree, elements[0], centroid);
+  dist = t8_vec_dist(adapt_data->midpoint, centroid);
```
```suggestion
  const dist = t8_vec_dist(adapt_data->midpoint, centroid);
```

Davknapp commented on example/figure_thesis/t8_is_family_fail.cxx:
```diff
@@ -0,0 +1,109 @@
+}
+
+
+int
```
Add documentation

Davknapp commented on example/figure_thesis/t8_is_family_fail.cxx:
```diff
@@ -0,0 +1,109 @@
+
+  struct t8_adapt_data adapt_data = { {0.5, 0.5, 0}, 0.3 };
+  
+  t8_forest_write_vtk (forest, "/home/ioannis/VBshare/paraview_export/t8_family_fail_base");
```
Please do not add hard coded paths to the repo. You can add them via a command-line option.

Davknapp commented on example/figure_thesis/t8_is_family_fail.cxx:
```diff
@@ -0,0 +1,109 @@
+  forest = t8_forest_new_uniform (cmesh, t8_scheme_new_default_cxx (), level, 0, comm);
+  T8_ASSERT (t8_forest_is_committed (forest));
+
+  struct t8_adapt_data adapt_data = { {0.5, 0.5, 0}, 0.3 };
```
I think it would be nice to add the data of the sphere via a command-line option. You increase the usability that way too.

Davknapp commented on example/figure_thesis/t8_is_family_fail.cxx:
```diff
@@ -0,0 +1,109 @@
+  t8_forest_write_vtk (forest, "/home/ioannis/VBshare/paraview_export/t8_family_fail_base");
+
+  forest = t8_forest_new_adapt (forest, t8_adapt_callback_remove, 0, 0, &adapt_data);
+  t8_forest_write_vtk (forest, "/home/ioannis/VBshare/paraview_export/t8_family_fail_remove");
```
NO hard coded paths!

Davknapp commented on example/figure_thesis/t8_is_family_fail.cxx:
```diff
@@ -0,0 +1,109 @@
+   * set forest->is_incomplete = 0; in t8_forest.c and 
+   * remove line 512ff in t8_forest_adapt.cxx */
+  forest = t8_forest_new_adapt (forest, t8_adapt_callback_coarse, 0, 0, &adapt_data);
+  t8_forest_write_vtk (forest, "/home/ioannis/VBshare/paraview_export/t8_family_fail_coare");
```
NO hard coded paths.

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+#include <t8_forest.h>
+#include <t8_schemes/t8_default/t8_default_cxx.hxx>
+#include <t8_vec.h>
+
```
Please add a description about what this examples does. My comments regarding documentations in the file `t8_is_family_fail` are also valid for this file! Also my comments regarding hard coded paths. Please have a look at the sc_options to add such optional parameters via the commandline. You increase the flexibility and reusability of your examples with only a little effort.

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+  const struct t8_adapt_data *adapt_data = (const struct t8_adapt_data *) t8_forest_get_user_data (forest);
+  
+  double  centroid[3];
+  double  dist;
```
```suggestion
```

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+  T8_ASSERT (adapt_data != NULL);
+
+  t8_forest_element_centroid (forest_from, which_tree, elements[0], centroid);
+  dist = t8_vec_dist(adapt_data->midpoint, centroid);
```
```suggestion
  const dist = t8_vec_dist(adapt_data->midpoint, centroid);
```

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+  const struct t8_adapt_data *adapt_data = (const struct t8_adapt_data *) t8_forest_get_user_data (forest);
+  
+  double  centroid[3];
+  double  dist;
```
```suggestion
```

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+  T8_ASSERT (adapt_data != NULL);
+
+  t8_forest_element_centroid (forest_from, which_tree, elements[0], centroid);
+  dist = t8_vec_dist(adapt_data->midpoint, centroid);
```
```suggestion
  const dist = t8_vec_dist(adapt_data->midpoint, centroid);
```

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+  const struct t8_adapt_data *adapt_data = (const struct t8_adapt_data *) t8_forest_get_user_data (forest);
+  
+  double  centroid[3];
+  double  dist;
```
See my above remarks regarding this variable.

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+  t8_cmesh_t          cmesh;
+  t8_forest_t         forest;
+
+  const int           level = 8;
```
This can be a commandline-parameter too.

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+  forest = t8_forest_new_uniform (cmesh, t8_scheme_new_default_cxx (), level, 0, comm);
+  T8_ASSERT (t8_forest_is_committed (forest));
+
+  struct t8_adapt_data adapt_data = { {0.5, 0.5, 0}, 0.3, 0.005 };
```
This should be a command-line parameter.

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+
+  forest = t8_balance (forest);
+
+  forest = t8_forest_new_adapt (forest, t8_adapt_callback_coarse, 0, 0, &adapt_data);
```
Do you intend to refine iteratively here? This is not the way we want to do that. You can add the maximum refinement level via `t8_forest_set_user_data` and call `t8_forest_set_adapt + t8_forest_commit` (or `t8_forest_new_adapt` should work as well).

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+  
+  t8_forest_write_vtk (forest, "/home/ioannis/VBshare/paraview_export/t8_example_0_base");
+
+  t8_debugf ("[IL] REFINE \n");
```
```suggestion
```
If this is only used for debugging purprose, it should be removed.

Davknapp commented on example/figure_thesis/t8_ring.cxx:
```diff
@@ -0,0 +1,177 @@
+
+  t8_forest_write_vtk (forest, "/home/ioannis/VBshare/paraview_export/t8_example_1_refine_balance");
+
+  t8_debugf ("[IL] REMOVE \n");
```
```suggestion
```

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+#include <t8_schemes/t8_default/t8_default_cxx.hxx>
+#include <t8_vec.h>
+#include "t8_cmesh/t8_cmesh_testcases.h"
+
```
Please add a description about this example. Furthermore, every function should be documented.

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+
+struct t8_adapt_data
+{
+  double  midpoint[6][3];
```
```suggestion
  double  midpoint[6][3];
  int num_spheres;
```

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+  const struct t8_adapt_data *adapt_data = (const struct t8_adapt_data *) t8_forest_get_user_data (forest);
+  
+  double  centroid[3];
+  double  dist;
```
This can be const (in the for-loop)

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+
+  t8_forest_element_centroid (forest_from, which_tree, elements[0], centroid);
+
+  for (int i = 0; i < 6; i++) {
```
```suggestion
  for (int i = 0; i < adapt_data->num_spheres; i++) {
```
Magic-numbers should be avoided.

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+  const struct t8_adapt_data *adapt_data = (const struct t8_adapt_data *) t8_forest_get_user_data (forest);
+  
+  double  centroid[3];
+  double  dist;
```
See comments above regarding `dist` and the magic-numbers in the for loop.

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+  const struct t8_adapt_data *adapt_data = (const struct t8_adapt_data *) t8_forest_get_user_data (forest);
+  
+  double  centroid[3];
+  double  dist;
```
See  comments above.

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+  const struct t8_adapt_data *adapt_data = (const struct t8_adapt_data *) t8_forest_get_user_data (forest);
+  
+  double  centroid[3];
+  double  dist;
```
See comments above.

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+  t8_cmesh_t          cmesh;
+  t8_forest_t         forest, forest_1, forest_2;
+
+  const int           level = 4;
```
This can be a command-line parameter.

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+
+  comm = sc_MPI_COMM_WORLD;
+
+  //cmesh = t8_cmesh_new_hypercube (T8_ECLASS_HEX, comm, 0, 0, 0);
```
Remove this comment, or make it a possible configuration of this example

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+
+  //cmesh = t8_cmesh_new_hypercube (T8_ECLASS_HEX, comm, 0, 0, 0);
+  cmesh = t8_cmesh_new_hypercube_hybrid (comm, 0, 0);
+  //cmesh = t8_test_create_cmesh (100); 
```
```suggestion
```

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+  int                 mpiret;
+  sc_MPI_Comm         comm;
+  t8_cmesh_t          cmesh;
+  t8_forest_t         forest, forest_1, forest_2;
```
These are bad variable-names. Please add more context.

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+  else {
+    t8_global_productionf("Forests are not equal.\n");
+  }
+  //for (int i = 0; i < level+1; i++)
```
Remove the comments

Davknapp commented on example/remove/t8_example_balls.cxx:
```diff
@@ -0,0 +1,225 @@
+  forest_1 = t8_forest_new_adapt (forest  , t8_adapt_callback_init  , 0, 0, &adapt_data);
+  forest_1 = t8_forest_new_adapt (forest_1, t8_adapt_callback_remove, 0, 0, &adapt_data);
+  
+  t8_forest_write_vtk (forest_1, "/home/ioannis/VBshare/paraview_export/t8_example_1");
```
NO hard coded paths!

Davknapp reviewed (commented):
This is a first batch of comments, regarding your examples. More will follow.
Some general feedback:
- Please stop adding your paths to the examples. Our examples should also run on machines other than yours. 
- Your examples can easily be improved by adding command-line options via sc_options. 
- The examples show the potential of your new features! Very nice!

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
 
-/* Check the lastly inserted elements of an array for recursive coarsening.
+#if T8_ENABLE_DEBUG
+/** Return zero if the first \a num_elemets in \a elements are not a (sub)family.
```
```suggestion
/** Return zero if the first \a num_elements in \a elements are not a (sub)family.
```

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+ *       not valid. 
+ */
+static int
+t8_forest_is_family_callback (t8_eclass_scheme_c *ts,
```
Can all input-parameters be const?

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+                              t8_element_t **elements) {
+  t8_element_t       *element_parent;
+  t8_element_t       *element_parent_compare;
+  int                 iter;
```
```suggestion
```

Davknapp replied:
Please declare iterator-variables in the for-loops

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+  t8_element_t       *element_parent_compare;
+  int                 iter;
+  int                 num_siblings;
+  int                 level;
```
```suggestion
```

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+    T8_ASSERT (ts->t8_element_is_valid (elements[iter]));
+  }
+
+  level = ts->t8_element_level(elements[0]);
```
```suggestion
 const int  level = ts->t8_element_level(elements[0]);
```

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+  t8_element_t       *element_parent;
+  t8_element_t       *element_parent_compare;
+  int                 iter;
+  int                 num_siblings;
```
```suggestion
```
Please have a look at our updated coding-guidelines: https://github.com/DLR-AMR/t8code/wiki/Coding-Guideline
It recently changed, but new additions should follow this guidelines.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+  ts->t8_element_new (1, &element_parent_compare);
+  ts->t8_element_new (1, &element_parent);
+  ts->t8_element_parent (elements[0], element_parent);
+  num_siblings = ts->t8_element_num_siblings (elements[0]);
```
```suggestion
 const int  num_siblings = ts->t8_element_num_siblings (elements[0]);
```

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+  ts->t8_element_parent (elements[0], element_parent);
+  num_siblings = ts->t8_element_num_siblings (elements[0]);
+
+  for (iter = 0; iter < num_elements; iter++) {
```
```suggestion
  for (int iter = 0; iter < num_elements; iter++) {
```

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+ * \param [in] tscheme       The element scheme for current local tree 
+ *                           where the elements are from.
+ * \param [in] elements      The elements array.
+ * \param [in] num_elements  The first \a num_elements to be checked in \a elements.
```
Add `\returns` to the doxygen comment

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+
+  if (num_elements < num_siblings) {
+    for (iter = num_elements; iter < num_elements; iter++) {
+      if (elements[iter] != NULL) {
```
Why is this check necessary? In line 54 all elements have been validated, if this part of the code is reached all elements are valid and not NULL. 
Please add more documentation to your implementations. It is not trivial, why we need these very similar loops.

Davknapp replied:
Is there an update?

ililikakis replied:
You are right, it is not necessary.

ililikakis replied:
The check is removed.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+               t8_element_array_t *telements,
+               t8_locidx_t telements_pos)
+{
+  t8_element_t       *element; /* element with index telements_pos */
```
Please update the constness and variable scope of these variables accroding to our coding guidelines. Check the types of the variables, too. In some cases you compare `int` with a `t8_locidx_t`

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+   * get coarsened recursively.
+   * If child_id is not last, elements cannot be coarsened recursively.
+   * Elements with level 0 cannot be further coarsened. */
+  if (!(child_id > 0 && child_id == num_siblings - 1) || level_current == 0) {
```
Do you check with `!(child_id > 0)` for the family-size? Child_id does not represent the size of the family, furthermore the first element of a family has child_id 0. Either update this check or the documentation.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+    T8_ASSERT (0 <= pos && pos < elements_in_array);
+    element_compare = t8_element_array_index_locidx (telements, pos);
+    /* Quick check by level. Not mandatory. */
+    level_compare = ts->t8_element_level (element_compare);
```
If this is not mandatory, it should be an assertion.

ililikakis replied:
The test is not complex and is intended to prevent parent elements from being calculated and compared too frequently.
In this way, computing time can probably be saved in most cases.
I will extend the documentation to make this clearer.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+    }
+  }
+
+  /* If the current family is smaller than possible, check that the first 
```
```suggestion
  /* If the current family is smaller than possible, check if the first 
```

Davknapp replied:
We want to avoid an overlapping, right?

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -22,14 +22,197 @@
+  t8_locidx_t         pos;
+  t8_locidx_t         el_iter; /* Loop running variable */
+  t8_locidx_t         elements_in_array;
+  t8_locidx_t         num_siblings;
```
`t8_element_num_siblings` returns an int. 
```suggestion
  int        num_siblings;
```

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -54,69 +238,82 @@ t8_forest_adapt_coarsen_recursive (t8_forest_t forest, t8_locidx_t ltreeid,
 {
   t8_element_t       *element;
```
Update variable declarations&scope to our current coding-guidelines.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -54,69 +238,82 @@ t8_forest_adapt_coarsen_recursive (t8_forest_t forest, t8_locidx_t ltreeid,
+      num_elements_to_adapt_callback = 1;
     }
-    if (isfamily
+    T8_ASSERT (forest->is_incomplete ? (!is_family || t8_forest_is_family_callback 
```
Please add documentation to this rather long assertion

Davknapp replied:
Why do we only check in the assertion if `fam` is a family if we pass an incomplete callback? Shouldn't we test that for incomplete forests, too? That way, we can skip the following if-condition.

ililikakis replied:
I do not understand the question and I don't see how to skip the if-condition. Also, what do you mean by incomplete forests? No trees can be removed.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -168,37 +369,51 @@ t8_forest_adapt_refine_recursive (t8_forest_t forest, t8_locidx_t ltreeid,
 void
 t8_forest_adapt (t8_forest_t forest)
 {
   t8_forest_t         forest_from;
```
Please update this according to our coding guidelines.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -257,16 +472,11 @@ t8_forest_adapt (t8_forest_t forest)
       tscheme->t8_element_num_siblings (first_element_from);
     /* Buffer for a family of new elements */
-    elements = T8_ALLOC (t8_element_t *, num_children);
+    elements           = T8_ALLOC (t8_element_t *, num_children);
```
This is not indented correctly. Is your indent-script up to date?

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -257,16 +472,11 @@ t8_forest_adapt (t8_forest_t forest)
+    elements           = T8_ALLOC (t8_element_t *, num_children);
     /* Buffer for a family of old elements */
-    elements_from = T8_ALLOC (t8_element_t *, curr_size_elements_from);
+    elements_from      = T8_ALLOC (t8_element_t *, curr_size_elements_from);
```
This is not indented correctly.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -294,41 +509,54 @@ t8_forest_adapt (t8_forest_t forest)
+       * The output will be  1 if the element should be refined
+       *                     0 if the element should remain as is
+       *                    -1 if we passed a family and it should get coarsened
+       *                    -2 if the element should be removed.
```
We start to produce to much magic-numbers here. Can you replace these numbers by an enum in a new branch and make a PR to the DLR-AMR main-branch?

ililikakis replied:
As discussed with Johannes, enums become problematic when transition cells get introduced.
Nevertheless, I took the time to play around a bit, see the feature-enum_set_adapt_fn_return branch.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -397,45 +626,85 @@ t8_forest_adapt (t8_forest_t forest)
+
     /* Check that if we had recursive adaptation, the refine list is now empty. */
     T8_ASSERT (!forest->set_adapt_recursive || refine_list->elem_count == 0);
+#if 1
```
This seems to switch between a functional and a work in progress implementation. Please decide for one version.
Having a look at the wip-implementation it all depends on t8_forest_remove_tree, which is not fully functional yet. If we want to implement the deletion of trees yet, we can discuss this function in more detail.

Davknapp commented on src/t8_forest/t8_forest_cxx.cxx:
```diff
@@ -41,6 +41,146 @@
+                                t8_element_t **elements,
+                                size_t elements_size)
+{
+  t8_tree_t           tree;
```
Please update the declaration of variables to our updated coding guidelines.

Davknapp commented on src/t8_forest/t8_forest_cxx.cxx:
```diff
@@ -41,6 +41,146 @@
 /* We want to export the whole implementation to be callable from "C" */
 T8_EXTERN_C_BEGIN ();
 
+int
```
I think, we should move this function into the element-scheme. For a general scheme this implementation is good, but for our default-scheme we should be able to implement an ìs_incomplete_family` check without any allocations. For all elements we assume to be in a reference cube. We could check, if the coordinates of all reference-cubes are consistent, if all levels are consistent. Since each element is uniquely defined, by coordinate, level and type we could then compute all child_ids and check if no child_ids is computed twice. All can be done in constant time and we avoid the allocation for two buffer elements. We can discuss this idea in our next meeting, if you agree or disagree.

Davknapp commented on src/t8_forest/t8_forest_cxx.cxx:
```diff
@@ -136,6 +276,49 @@ t8_forest_min_nonempty_level (t8_cmesh_t cmesh, t8_scheme_cxx_t *scheme)
+int
+t8_forest_no_overlap (t8_forest_t forest) {
+#if T8_ENABLE_DEBUG
+  t8_locidx_t         num_local_trees;
```
Please update the variables to our coding guideline.

Davknapp commented on src/t8_forest/t8_forest_cxx.cxx:
```diff
@@ -136,6 +276,49 @@ t8_forest_min_nonempty_level (t8_cmesh_t cmesh, t8_scheme_cxx_t *scheme)
+    ts->t8_element_new(1, &element_nca);
+    /* Iterate over all elements in current tree */
+    for (ielem = 0; ielem < elems_in_tree-1; ielem++) {
+      /* Compare each two consecutive elements. If one element is
```
Please update this comment. You only check the level in line 309, which is enough in this case, nevertheless a `t8_element_compare` is expected following this comment.

Davknapp commented on src/t8_forest/t8_forest_cxx.cxx:
```diff
@@ -1372,55 +1555,127 @@ t8_forest_populate (t8_forest_t forest)
+  T8_ASSERT (forest != NULL);
+  T8_ASSERT (forest->first_local_tree > -1);
+  T8_ASSERT (forest->first_local_tree < forest->global_num_trees);  
+  T8_ASSERT ((forest->first_local_tree == 0 && forest->last_local_tree == -1) ?
```
Please add documentation to non-trivial assertions.
Can you split this assertion into multiple assertions? We prefer readable code over super-compact code.

Davknapp commented on src/t8_forest/t8_forest_cxx.cxx:
```diff
@@ -1372,55 +1555,127 @@ t8_forest_populate (t8_forest_t forest)
+        return 0;
+      }
+      T8_ASSERT (global_neighbour_tree_idx < forest->global_num_trees); 
+      T8_ASSERT (global_neighbour_tree_idx > -1 ?
```
Please document this non-trivial assertion and, if possible, split it into multiple ones.

Davknapp commented on src/t8_forest/t8_forest_cxx.cxx:
```diff
@@ -1372,55 +1555,127 @@ t8_forest_populate (t8_forest_t forest)
+                  global_neighbour_tree_idx == -1);
+    }
+    else {
+      SC_ABORT ("Method t8_forest_last_tree_shared is not implemented.\n");
```
Shouldn't this be handled via t8_forest_tree_shared(forest, 1)? Please specify this SC_ABORT.

Davknapp commented on src/t8_forest/t8_forest_cxx.cxx:
```diff
@@ -1492,7 +1747,88 @@ t8_forest_copy_trees (t8_forest_t forest, t8_forest_t from, int copy_elements)
-/* Search for a linear element id (at forest->maxlevel) in a sorted array of
+/* TODO: To work for many processes, global_num_trees, tree_offsets,
+ * first_local_tree and last_local_tree must be calculated for each process. */
+/* TODO: Extend to nonempty trees. */
```
This looks like a rather unfinished function. Either implement it fully, oder move it to another branch or delete it.

Davknapp commented on src/t8_forest/t8_forest.c:
```diff
@@ -57,6 +57,8 @@ t8_forest_init (t8_forest_t *pforest)
   forest->set_balance = -1;
   forest->maxlevel_existing = -1;
   forest->stats_computed = 0;
+  /* TODO [IL] Set init to 0 if tests are done. */
```
Are the tests done? A forest should be complete by default.

Davknapp commented on src/t8_forest/t8_forest_iterate.cxx:
```diff
@@ -394,14 +394,22 @@ t8_forest_iterate_replace (t8_forest_t forest_new,
                            t8_forest_t forest_old,
                            t8_forest_replace_t replace_fn)
 {
```
Please update to our new coding-guidelines.

Davknapp commented on src/t8_forest/t8_forest_iterate.cxx:
```diff
@@ -420,15 +428,134 @@ t8_forest_iterate_replace (t8_forest_t forest_new,
+
+    /* Refine flag of -3 is not defined. Its set here for testing reasons.
+     * later in */
+    refine = -3;
```
Please avoid magic-numbers. Do we use it in a test? 
Please see my above comment regarding refinement/coarsen/delete-values. This parameter should be added. If this parameter is not needed anymore, remove it.

Davknapp commented on src/t8_forest/t8_forest_iterate.cxx:
```diff
@@ -420,15 +428,134 @@ t8_forest_iterate_replace (t8_forest_t forest_new,
+                  "Family is not complete");
+          }
+#endif
+          t8_debugf("[IL] itree %i, elem_old %i, elem_new %i, refine %i, family_size 1\n", 
```
Please move it into the debug-makros, or make it proper-debugging output (Remove the [IL] at the beginning)

Davknapp commented on src/t8_forest/t8_forest_iterate.cxx:
```diff
@@ -420,15 +428,134 @@ t8_forest_iterate_replace (t8_forest_t forest_new,
+                "Element is not the first of the family.");
+          }
+#endif
+          t8_debugf("[IL] itree %i, elem_old %i, elem_new %i, refine %i, family_size %i\n", 
```
See comment above.

Davknapp commented on src/t8_forest/t8_forest_iterate.cxx:
```diff
@@ -420,15 +428,134 @@ t8_forest_iterate_replace (t8_forest_t forest_new,
+        if (!ts->t8_element_compare (elem_new, elem_old)) {
+          /* elem_new = elem_old */
+          refine = 0;
+          t8_debugf("[IL] itree %i, elem_old %i, elem_new %i, refine %i, family_size 1\n", 
```
See comment above.

Davknapp commented on src/t8_forest/t8_forest_iterate.cxx:
```diff
@@ -420,15 +428,134 @@ t8_forest_iterate_replace (t8_forest_t forest_new,
+      }
+      if (refine == -2) {
+        /* element got removed */
+        t8_debugf("[IL] itree %i, elem_old %i, elem_new %i, refine %i, family_size 1\n", 
```
See comment above.

Davknapp reviewed (commented):
A second chunk of comments! Nice work in general, please address my comments.

Davknapp commented on benchmarks/t8_time_fractal.cxx:
```diff
@@ -0,0 +1,465 @@
+#include <sc_flops.h>
+#include <sc_statistics.h>
+#include <sc_options.h>
+
```
Please document the file, the functions and update the variables to our coding guidelines

Davknapp commented on benchmarks/t8_time_fractal.cxx:
```diff
@@ -0,0 +1,465 @@
+
+  /* initialize command line argument parser */
+  opt = sc_options_new (argv[0]);
+  sc_options_add_switch (opt, 'h', "help", &helpme,
```
Please add the default values in the description of the parameters.

Davknapp commented on test/Makefile.am:
```diff
@@ -25,8 +25,11 @@ t8code_test_programs = \
     test/t8_test_netcdf_linkage \
-    test/t8_test_vtk_linkage
-
+    test/t8_test_vtk_linkage \
```
The name of the test programs and sources do not match. Please update.
Additionally all new tests should be googletests.

Davknapp commented on benchmarks/t8_time_fractal.cxx:
```diff
@@ -0,0 +1,465 @@
+  char                usage[BUFSIZ];
+  char                help[BUFSIZ];
+  int                 initial_level = 0;
+  int                 end_level_it = 4;
```
It is confusing to have two different types of final levels. Please use only one final level and add a switch to change between iterative and recursive refinement.

Davknapp commented on example/remove/t8_example_moving_circle.cxx:
```diff
@@ -0,0 +1,208 @@
+#include <t8_schemes/t8_default/t8_default_cxx.hxx>
+#include <t8_vec.h>
+#include <string>
+
```
Please document the file, all functions and update the variables to our coding guidelines.

Davknapp replied:
All examples in this folder are very similar. In my opinion we do not need 3 very similar examples. If you want store them for reproducability, add them to t8_data. Pick the example that shows the new functionality best (or adapt it a little such that one shows the functionality of your additions best).

Davknapp commented on example/remove/t8_example_ring.cxx:
```diff
@@ -0,0 +1,139 @@
+#include <t8_forest.h>
+#include <t8_schemes/t8_default/t8_default_cxx.hxx>
+#include <t8_vec.h>
+
```
See my comments above. Pick one remove-example and polish it :)

Davknapp commented on src/t8_forest.h:
```diff
@@ -71,8 +71,8 @@ T8_EXTERN_C_BEGIN ();
  * \param [in] which_tree      The local tree containing \a outgoing and \a incoming
```
```suggestion
 * \param [in] which_tree      The local tree containing \a first_outgoing and \a first_incoming
```

Davknapp commented on src/t8_forest.h:
```diff
@@ -83,8 +83,10 @@ T8_EXTERN_C_BEGIN ();
+ * the number of family members and \a num_incoming will be 1. 
+ * If an element is being removed, \a refine and \a num_outgoing will be 1 and 
+ * \a num_incoming will be 0. 
+ * Else \a refine will be 0 and \a num_outgoing and \a num_incoming will both be 1.
```
Can you also update the parameters refine, num_outgoing and num_incoming to const?

Davknapp commented on src/t8_forest.h:
```diff
@@ -161,6 +164,14 @@ int                 t8_forest_is_initialized (t8_forest_t forest);
+/** Check whether the forest has overlapping elements.
+ * \param [in] forest_a The forest to consider.
+ * \return              True if \a forest has no elements which are inside each other.
+ * \note This function is not collective. It only returns the state on the current
```
I think it would make sense to make this function collective. Otherwise the communication if the forest is overlapping somewhere has to be rewritten in every application and we want to avoid code duplication.

ililikakis replied:
Done. But it still only checks local overlap for each process. See new documentation for more details.
I can merge t8_forest_no_overlap and t8_forest_partition_test_boundery_element functions to check for global overlap.

Davknapp commented on src/t8_forest/t8_forest_partition.cxx:
```diff
@@ -162,6 +162,65 @@ t8_forest_partition_test_desc (t8_forest_t forest)
 
+void
+t8_forest_partition_test_boundery_element (t8_forest_t forest)
+{
```
Please update the variables to our coding guidelines.

Davknapp commented on src/t8_forest/t8_forest_partition.h:
```diff
@@ -78,6 +78,15 @@ void                t8_forest_partition_data (t8_forest_t forest_from,
                                               const sc_array_t *data_in,
                                               sc_array_t *data_out);
 
+/** Test if last descendant of the last element of current rank has
```
```suggestion
/** Test if the last descendant of the last element of current rank has
```

Davknapp commented on src/t8_forest/t8_forest_private.h:
```diff
@@ -37,6 +37,31 @@ T8_EXTERN_C_BEGIN ();
+ *                  family member, return 0. Therefore, if \return is x > 0, 
+ *                  the first x elements in \a elements form a family.
+ */
+int                 t8_forest_is_incomplete_family (t8_forest_t forest,
```
No input-parameter is changed by the function, so they should all be const. As said befor, please move this function to the common functions of the default scheme.

Davknapp commented on src/t8_forest/t8_forest_private.h:
```diff
@@ -107,11 +132,18 @@ void                t8_forest_copy_trees (t8_forest_t forest,
                                           int copy_elements);
 
+/** Delete a local empty tree from a forest.
+ * \param [in]  forest    The forest.
```
```suggestion
 * \param [in, out]  forest    The forest.
```

Davknapp replied:
We change the forest in this function.

Davknapp commented on src/t8_forest/t8_forest_private.h:
```diff
@@ -107,11 +132,18 @@ void                t8_forest_copy_trees (t8_forest_t forest,
+ * \param [in]  forest    The forest.
+ * \param [in]  ltree_id  The local id of the tree in \a forest to be deleted.
+ */
+void                t8_forest_remove_tree (t8_forest_t forest,
```
Please only add fully functional functions.

Davknapp commented on src/t8_forest/t8_forest_private.h:
```diff
@@ -107,11 +132,18 @@ void                t8_forest_copy_trees (t8_forest_t forest,
+ * \param [in]  ltree_id  The local id of the tree in \a forest to be deleted.
+ */
+void                t8_forest_remove_tree (t8_forest_t forest,
+                                           t8_locidx_t ltree_id);
```
```suggestion
                                           const t8_locidx_t ltree_id);
```

Davknapp commented on src/t8_forest/t8_forest_types.h:
```diff
@@ -81,6 +81,8 @@ typedef struct t8_forest
   int                 maxlevel_existing;/**< If >= 0, the maximum occurring refinemnent level of a forest element. */
   int                 do_dup;           /**< Communicator shall be duped. */
   int                 dimension;        /**< Dimension inferred from \b cmesh. */
+  int                 is_incomplete;    /**< Flag to decide whether the forest ist complete or not.
```
Please be more specific, what does incomplete/complete mean?

ililikakis replied:
Now it should be fine.

Davknapp commented on test/t8_forest_incomplete/t8_test_remove_rand.cxx:
```diff
@@ -0,0 +1,199 @@
+#include "t8_cmesh/t8_cmesh_testcases.h"
+#include <t8_forest/t8_forest_partition.h>
+#include <t8_forest/t8_forest_types.h>
+
```
Please document this file, all functions and update the variables to our coding guidelines. Please make this a parametrized google-test.

ililikakis replied:
I have removed this test because it is based on randomness. Instead I created the test `t8_gtest_permute_hole.cxx`.

Davknapp commented on test/t8_forest_incomplete/t8_test_recursive.cxx:
```diff
@@ -0,0 +1,225 @@
+  along with t8code; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
```
Please document the file, all functions and update the variables to our coding guidelines. This should be a parametrized google test. If you want to test only for one rank, use Gtest_skip.

ililikakis replied:
Done.

Davknapp commented on test/t8_forest_incomplete/t8_test_elements_remove.cxx:
```diff
@@ -0,0 +1,289 @@
+  along with t8code; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
```
Please document this file, all functions and update all variables to our coding guidelines. 
This Test should be a parametrized google-test.

So far all your tests are rather complex, which is good, but we should also add a very simple basic test, which will make it easier to debug in the future. I suggest to construct a single element, refine it once and then delete between 0 and `num_children-1` elements. That way we can:
- check if the elements are deleted
- also check `is_incomplete_family`

ililikakis replied:
Done. I have removed this complex test. See `t8_gtest_permute_hole.cxx` instead.

Davknapp commented on test/t8_forest_incomplete/t8_test_elements_interpolation.cxx:
```diff
@@ -0,0 +1,309 @@
+#include "t8_cmesh/t8_cmesh_testcases.h"
+
+#include <t8_forest/t8_forest_partition.h>
+
```
This is more of a visual test of the interpolation function and therefore should be moved to the examples folder. Can we implement a test that checks the correctness of the interpolation function?

Furthermore, please do not add your paths to the repository!

ililikakis replied:
Done. See `t8_gtest_iterate_replace.cxx` instead.

ililikakis committed:
removed example files for thesis

ililikakis committed:
Merge branch 'main' into feature-remove-elements

ililikakis committed:
fixed directory in test

ililikakis committed:
t8_example_balls to _spheres

ililikakis committed:
Update src/t8_forest/t8_forest_adapt.cxx

Co-authored-by: David Knapp <david.knapp@dlr.de>

ililikakis committed:
changed scope of variables

ililikakis committed:
Update src/t8_forest/t8_forest_adapt.cxx

Co-authored-by: David Knapp <david.knapp@dlr.de>

ililikakis committed:
Update src/t8_forest/t8_forest_adapt.cxx

Co-authored-by: David Knapp <david.knapp@dlr.de>

ililikakis committed:
add return to documentation of t8_forest_is_family_callback

ililikakis committed:
update scope of variable in  t8_forest_is_family_callback

ililikakis committed:
update documentation in t8_forest_pos

ililikakis committed:
update constness and variable scope in t8_forest_pos

ililikakis committed:
update variable types in t8_forest_pos

ililikakis committed:
fixed documentation in t8_forest_pos

ililikakis committed:
reduced scope of variables in t8_forest_adapt_coarsen_recursive

ililikakis committed:
fixed variable type issues in t8_forest_adapt_coarsen_recursive

ililikakis committed:
reduced scope of variables in t8_forest_adapt_refine_recursive

ililikakis committed:
removed unnecessary if-condition in t8_forest_adapt_refine_recursive

ililikakis committed:
changed num_elements_to_adapt_callback into int in t8_forest_adapt_coarsen_recursive

ililikakis committed:
indent t8_forest_adapt.cxx

ililikakis committed:
indent example/remove/t8_example_spheres.cxx

ililikakis committed:
reduced scope of variables in t8_forest_is_incomplete_family

ililikakis committed:
reduced scope of variables in t8_forest_no_overlap

ililikakis committed:
indent t8_forest_cxx.cxx

ililikakis committed:
extended documentation of t8_forest_no_overlap

ililikakis committed:
removed function t8_forest_remove_tree

ililikakis committed:
update scope of variables in t8_forest_iterate_replace

ililikakis committed:
Merge branch 'main' into feature-remove-elements

ililikakis committed:
removed magic number and changed for into while loop

ililikakis committed:
removed personal debuging prints

ililikakis committed:
const input of t8_forest_replace

ililikakis committed:
extend documentation in t8_forest_iterate_replace

ililikakis committed:
indent t8_forest_iterate_replace

ililikakis committed:
reduced magic numbers in t8_forest_iterate_replace

ililikakis committed:
reduced sscope of elem_parent in t8_forest_iterate_replace

ililikakis committed:
made t8_forest_no_overlap a collactive function

ililikakis committed:
made input const of t8_forest_is_family_callback

ililikakis committed:
simplified T8_ASSERTs

ililikakis committed:
indent t8_forest_adapt.cxx

ililikakis committed:
simplified T8_ASSERTs

ililikakis committed:
indent t8_forest_cxx.cxx

ililikakis committed:
fixed input issue in t8_forest_is_family_callback

ililikakis committed:
extend SC_ABORT comments in t8_forest_tree_shared

ililikakis committed:
set complete trees as default

ililikakis committed:
indent t8_forest_adapt.cxx

ililikakis committed:
extended documentation

ililikakis committed:
const input forest in t8_forest_is_incomplete_family

ililikakis committed:
applied coding guidelines to t8_forest_partition_test_boundery_element

ililikakis committed:
indent t8_forest_partition_test_boundery_element

ililikakis committed:
const input of t8_forest_partition_test_boundery_element

ililikakis committed:
fixed ASSSERT issue with incomplete trees in t8_forest_commit

ililikakis committed:
added ASSERT in t8_forest_adapt

ililikakis committed:
Merge branch 'main' into feature-remove-elements

ililikakis committed:
fixed MPI issue in t8_forest_no_overlap

ililikakis committed:
fixed is_incomplete set flag issue

ililikakis committed:
updated benchmark t8_time_fractal

ililikakis committed:
update t8_example_spheres.cxx

ililikakis committed:
removed similar examples

ililikakis committed:
removed similar examples - Makefile

ililikakis committed:
update t8_test_recursive.cxx

ililikakis committed:
update t8_test_recursive.cxx

ililikakis committed:
changed flag is_incomplete in incomplete_trees

ililikakis committed:
added first version of gauss_blob example

Davknapp commented on benchmarks/t8_time_fractal.cxx:
```diff
@@ -87,314 +164,275 @@ t8_adapt_menger (t8_forest_t forest,
+  /* Set up userdata to adapt forest. 
+   * user_data[0] -> level_end
+   * user_data[1] -> {0, -2} -> return for adapt callback */
+  int                 user_data[2] = { level_end, remove };
+  if (remove == 1) {
+    user_data[1] = -2;
+  }
```
```suggestion
  int                 user_data[2] = { level_end, (remove == 1) ? -2 : remove };
```

ililikakis replied:
done

Davknapp commented on benchmarks/t8_time_fractal.cxx:
```diff
@@ -87,314 +164,275 @@ t8_adapt_menger (t8_forest_t forest,
+   * greater 0, such that the refinement pattern can be applied. */
+  if (eclass == T8_ECLASS_HEX || eclass == T8_ECLASS_QUAD) {
+    if (level_initial < 2) {
+      level_initial = 2;
```
We should not change the level without the users permission. Please add an output for the user that:
a) the initial level has been changed.
b) that the function does not work on such small levels and abort the call of the function.

ililikakis replied:
Done

Davknapp commented on benchmarks/t8_time_fractal.cxx:
```diff
@@ -87,314 +164,275 @@ t8_adapt_menger (t8_forest_t forest,
-          t8_forest_set_adapt (forest_adapt, forest, t8_adapt_sierpinski_pyramid, 1);
-        }
+      t8_forest_set_user_data (forest_adapt, &user_data);
+      if (eclass == T8_ECLASS_QUAD) {
```
You can use a ```switch(eclass)``` here

ililikakis replied:
Done

Davknapp commented on example/remove/t8_example_gauss_blob.cxx:
```diff
@@ -0,0 +1,318 @@
+{
+  t8_vtk_data_field_t vtk_data;
+  vtk_data.type = T8_VTK_SCALAR;
+  strcpy (vtk_data.description, "Gauss");
```
It is better and safer to use a `snprintf` here.

ililikakis replied:
Done

Davknapp commented on example/remove/t8_example_spheres.cxx:
```diff
@@ -0,0 +1,241 @@
+  const int           num_spheres;
+  const double        spheres_radius_inner;
+  const double        spheres_radius_outer;
+  const double        midpoint[6][3];
```
This struct needs some adaptation. num_spheres is variable, but midpoint cannot store more than six midpoints. Please make it a dynamic array. 
Furthermore all spheres have the same inner and outer radius, which is fine for the moment but could be changed, too.

ililikakis replied:
The struct takes now a pointer for the midpoints.

Davknapp commented on src/t8_forest/t8_forest_iterate.cxx:
```diff
@@ -394,202 +394,205 @@ t8_forest_iterate_replace (t8_forest_t forest_new,
           /* element got removed */
-          refine = -2;
+          const int           refine = -2;
+          replace_fn (forest_old, forest_new, itree, ts, refine,
```
When I compile the code I get a warning:```
/t8code/src/t8_forest/t8_forest_iterate.cxx:549:40: warning: converting to non-pointer type 't8_locidx_t' {aka 'int'} from NULL [-Wconversion-null]
  549 |                       1, ielem_old, 0, NULL);```
  please have a look at it!

ililikakis replied:
Done. There should be no warning any more.

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -110,46 +107,40 @@ t8_forest_is_family_callback (t8_eclass_scheme_c *ts,
-   * Elements with level 0 cannot be further coarsened. */
-  if (!(child_id > 0 && child_id == num_siblings - 1) || level_current == 0) {
-    return INT32_MIN;
+  const size_t        elements_in_array =
+    t8_element_array_get_count (telements);
```
```suggestion
  const t8_locidx_t        elements_in_array =
    (t8_locidx_t )t8_element_array_get_count (telements);
```

ililikakis replied:
Done

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -158,13 +149,18 @@ t8_forest_pos (t8_forest_t forest,
+       && (size_t) el_iter < elements_in_array; el_iter++) {
     pos = telements_pos - el_iter;
-    T8_ASSERT (0 <= pos && pos < elements_in_array);
+    T8_ASSERT (0 <= pos && (size_t) pos < elements_in_array);
```
```suggestion
    T8_ASSERT (0 <= pos && pos < elements_in_array);
```

ililikakis replied:
Done

Davknapp commented on src/t8_forest/t8_forest_adapt.cxx:
```diff
@@ -406,15 +419,16 @@ t8_forest_adapt (t8_forest_t forest)
-  size_t              num_children; 
-  size_t              num_siblings; 
-  size_t              zz; 
+  size_t              num_children;
```
The types of these variables and the return-type of the element_functions do not match. Please use the same types.

ililikakis replied:
Should be fine now.

Davknapp commented on test/t8_forest_incomplete/t8_test_recursive.cxx:
```diff
@@ -136,90 +108,67 @@ t8_adapt_forest (t8_forest_t forest_from,
   return 0;
-}
\ No newline at end of file
+}
```
Please make this test a googletest

ililikakis replied:
Done

Davknapp reviewed (changes requested):
Thank you for your work, Another round of feedback! We are closing in onto the finish line ;)

ililikakis committed:
added vtkDataSet to vtkPointSet

ililikakis committed:
removed not required includes

ililikakis committed:
moved vtk files into new t8_vtk folder

ililikakis committed:
fixed header and makefile due to moving files

ililikakis committed:
moving functions

ililikakis committed:
rename function

ililikakis committed:
indent

ililikakis committed:
undo add vtkDataSet to vtkPointSet

Commit:
merged from main

Davknapp committed:
Merge remote-tracking branch 'origin/main' into cmesh_vtk_DataSet_translator_ext

ililikakis committed:
merge from main

ililikakis committed:
fixed unresolved mergeconflicts

ililikakis committed:
fixed include forest.h

ililikakis committed:
replaced tests

ililikakis committed:
transformed test into gtest

ililikakis committed:
changed range in t8_gtest_permute_hole.cxx

ililikakis committed:
add documentation

ililikakis committed:
indent

ililikakis committed:
add documentation

ililikakis committed:
fixed variable issue

ililikakis committed:
removed headerfiles

ililikakis committed:
added ASSER_TRUE

ililikakis committed:
fixed issue for debugging

ililikakis committed:
added test for iterate_replace

ililikakis committed:
set startlevel to 1

ililikakis committed:
fixed issue when removing last elements of forest

ililikakis committed:
add test - empty trees

ililikakis committed:
removed old assert

ililikakis committed:
support adapting forest with empty trees

ililikakis committed:
updated tests

ililikakis committed:
extended tests

ililikakis committed:
merge from main

ililikakis committed:
cleanup and documentation

ililikakis committed:
removed not necessary check

ililikakis committed:
shorten if-else condition

ililikakis committed:
if-else to switch-case

ililikakis committed:
added error, if levels are set wrong

ililikakis committed:
changed strcpy into snprintf

ililikakis committed:
removed static array

ililikakis committed:
fixed warning by passing -1 insted NULL to t8_forest_replace_t func

ililikakis committed:
changed datatype to avoid casting

ililikakis committed:
fixed datatype cast issues

ililikakis committed:
fixed ref issue

ililikakis committed:
fixed pointer issue

ililikakis committed:
fixed naming and scope

ililikakis committed:
docu

ililikakis committed:
fixed pointer issue

ililikakis committed:
skip some tests until empty global trees are supported

ililikakis committed:
fix skip

ililikakis committed:
fix unref/free issue

benegee committed:
switched to t8_forest_write_vtk_ext to be able to write ghost cells

benegee committed:
added BSD statement

ililikakis committed:
partition tests

ililikakis committed:
fixed scope in t8_forest_partition_fill_buffer(*)

ililikakis committed:
removed old assert

benegee committed:
Disable VK API

Since the export of ghost elements is not supported with the VTK API, we have to disable it

Co-authored-by: Sandro Elsweijer <49643115+sandro-elsweijer@users.noreply.github.com>

ililikakis committed:
removed debug prints

ililikakis committed:
fixed unref issue and reduced size

ililikakis committed:
indent

ililikakis committed:
extended test

Davknapp committed:
Merge remote-tracking branch 'origin/main' into cmesh_vtk_DataSet_translator_ext

Davknapp committed:
Merge branch 'cmesh_vtk_DataSet_translator_ext' of https://github.com/DLR-AMR/t8code into cmesh_vtk_DataSet_translator_ext

ililikakis committed:
added new example for empty trees

Davknapp committed:
Merge pull request #539 from DLR-AMR/cmesh_vtk_DataSet_translator_ext

Cmesh vtk data set translator [vtk_to_cmesh 6a/6]

benegee committed:
ran indent

benegee committed:
Merge branch 'ghost-cell-output-in-tutorial-step4' of github.com:bgeihe/t8code into ghost-cell-output-in-tutorial-step4

sandro-elsweijer committed:
Merge pull request #583 from bgeihe/ghost-cell-output-in-tutorial-step4

Enable ghost cell output in tutorial step 4

ililikakis committed:
indent

ililikakis committed:
Merge branch 'main' into feature-remove-elements

ililikakis committed:
slim down code t8_example_gauss_blob

ililikakis committed:
fixed uninitialized Werror

ililikakis committed:
fixed Werror

ililikakis committed:
false ASSERT

ililikakis committed:
MPI_INT to sc_MPI_INT

ililikakis committed:
MPI_LONG_LONG_INT to sc_MPI_LONG_LONG_INT

ililikakis committed:
set tree ids to t8_gloidx_t

ililikakis committed:
changed naming of variable

ililikakis committed:
chaged naming of variables

ililikakis committed:
indent
